airdrop/airdrop.gno

```diff
@@ -0,0 +1,215 @@
+package airdrop
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"std"
+	"time"
+
+	"gno.land/p/demo/grc/grc20"
+	i256 "gno.land/p/demo/int256"
+	"gno.land/p/demo/merkle"
+	u256 "gno.land/p/demo/uint256"
+)
+
+var (
+	ErrInvalidProof              = errors.New("INVALID_PROOF")
+	ErrInvalidClaimsCount        = errors.New("INVALID_CLAIMS_COUNT")
+	ErrFirstClaimIsNot128        = errors.New("FIRST_CLAIM_MUST_BE_MULT_128")
+	ErrRefundableTimestampNotSet = errors.New("REFUNDABLE_TIMESTAMP_NOT_SET")
+	ErrTooEarlyToRefund          = errors.New("TOO_EARLY_TO_REFUND")
+)
+
+type Claim struct {
+	ID      uint64
+	Claimee string
+	Amount  *u256.Uint // ? u256
+}
+
+func (c Claim) Bytes() []byte {
+	var buf bytes.Buffer
+	idBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(idBytes, c.ID)
+	buf.Write(idBytes)
+	buf.Write([]byte(c.Claimee))
+	amountStr := c.Amount.ToString()
+	buf.Write([]byte(amountStr))
+	return buf.Bytes()
+}
+
+type Config struct {
+	Root                [32]byte
+	RefundableTimestamp uint64 // TODO: timelock
+	RefundTo            std.Address
+}
+
+type Airdrop struct {
+	token         grc20.IGRC20
+	config        Config
+	claimedBitmap map[uint64]*u256.Uint
+	merkleTree    *merkle.Tree
```

Comment:
merkle.Tree exposed from p/demo/merkle is the whole tree containing all the nodes. For airdrop, we might not need to store the tree itself, as we only need to verify the user's submitted proofs against the stored root. We can call merkle.Verify(a.root, claim, proofs) instead.

Reply:
Yes, Including the entire tree consumes unnecessay space, but it was inevitable to allocated a field of the merkle tree to make easy to use. Currently, the tree has been removed and I'll explore other mthods (as it was awkward to use).